metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java

```diff
@@ -0,0 +1,454 @@
+        // Entries are sorted by resource type first, then REVERSE resource name.
+        // Therefore, we can find all the applicable ACLs by starting at
+        // (resource_type, resource_name) and stepping forwards until we reach an ACL with
+        // a resource name which is not a prefix of the current one.
+        //
+        // For example, when trying to authorize a TOPIC resource named foobar, we would
+        // start at element 2, and continue on to 3 and 4 following map:
+        //
+        // 1. rs=TOPIC rn=gar pt=PREFIX
+        // 2. rs=TOPIC rn=foobar pt=PREFIX
+        // 3. rs=TOPIC rn=foob pt=LITERAL
+        // 4. rs=TOPIC rn=foo pt=PREFIX
+        // 5. rs=TOPIC rn=eeee pt=LITERAL
+        //
+        // Once we reached element 5, we would stop scanning.
+        AuthorizationResultBuilder builder = new AuthorizationResultBuilder();
+        StandardAcl exemplar = new StandardAcl(
+            action.resourcePattern().resourceType(),
+            action.resourcePattern().name(),
+            PatternType.UNKNOWN,
+            "",
+            "",
+            AclOperation.UNKNOWN,
+            AclPermissionType.UNKNOWN);
+        checkSection(action, exemplar, requestContext, builder);
+        if (builder.foundDeny) return DENIED;
+
+        // In addition to ACLs for this specific resource name, there can also be wildcard
+        // ACLs that match any resource name. These are stored as type = LITERAL,
+        // name = "*". We search these next.
+        exemplar = new StandardAcl(
+            action.resourcePattern().resourceType(),
+            WILDCARD,
+            LITERAL,
+            "",
+            "",
+            AclOperation.UNKNOWN,
+            AclPermissionType.UNKNOWN);
+        checkSection(action, exemplar, requestContext, builder);
+        if (builder.foundDeny) return DENIED;
+
+        // If we found ALLOW ACLs, the action is allowed.
+        if (builder.foundAllow) {
+            if (log.isTraceEnabled()) {
+                log.trace("authorize(requestContext=" + requestContext + ", action=" +
+                    action + "): ALLOWED");
+            }
+            return ALLOWED;
+        }
+
+        // If nothing matched, we return the default result.
+        if (log.isTraceEnabled()) {
+            log.trace("authorize(requestContext=" + requestContext + ", action=" +
+                action + "): returning default result " + defaultResult);
+        }
+        return defaultResult;
+    }
+
+    void checkSection(Action action,
+                      StandardAcl exemplar,
+                      AuthorizableRequestContext requestContext,
+                      AuthorizationResultBuilder builder) {
+        NavigableSet<StandardAcl> headSet = aclsByResource.tailSet(exemplar, true);
+        String resourceName = action.resourcePattern().name();
+        for (Iterator<StandardAcl> iterator = headSet.iterator();
```

Comment:
I really prefer to be explicit here, since we are iterating over a set that is potentially changing. I know that Java will only evaluate the y in `for (x: y) {}` once but I think there are some languages that don't, so no point being obscure.